homeassistant/components/rointe/sensor.py

```diff
@@ -0,0 +1,160 @@
+        ROINTE_COORDINATOR
+    ]
+
+    # Hook a callback for discovered entities for the sensor entities.
+    coordinator.add_entities_for_seen_keys(
+        async_add_entities,
+        [RointeHaSensor, RointeEnergySensor, RointePowerSensor],
+        "sensor",
+    )
+
+
+class RointeHaSensor(RointeRadiatorEntity, SensorEntity):
+    """Sensor for the radiator current temperature."""
+
+    def __init__(
+        self,
+        radiator: RointeDevice,
+        coordinator: RointeDataUpdateCoordinator,
+    ) -> None:
+        """Init the sensor entity."""
+
+        description = SensorEntityDescription(
+            key="current_temperature",
+            name="Current Temperature",
+            native_unit_of_measurement=TEMP_CELSIUS,
+            device_class=SensorDeviceClass.TEMPERATURE,
+            state_class=SensorStateClass.MEASUREMENT,
+            entity_category=EntityCategory.DIAGNOSTIC,
+        )
+
+        # Set the name and ID of this sensor to be the radiator name/id and a prefix.
+        super().__init__(
+            coordinator,
+            radiator,
+            name=f"{radiator.name} {description.name}",
+            unique_id=f"{radiator.id}-{description.key}",
+        )
+
+        self.entity_description = description
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the current sensor value (Probe value)."""
+        return self._radiator.temp_probe
+
+
+class RointeEnergySensor(RointeRadiatorEntity, SensorEntity):
+    """Sensor for the radiator energy usage in Kw/h."""
+
+    def __init__(
+        self,
+        radiator: RointeDevice,
+        coordinator: RointeDataUpdateCoordinator,
+    ) -> None:
+        """Init the sensor entity."""
+
+        description = SensorEntityDescription(
+            key="energy",
+            name="Energy Consumption",
+            native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+            device_class=SensorDeviceClass.ENERGY,
+            state_class=SensorStateClass.TOTAL,
+            entity_category=EntityCategory.DIAGNOSTIC,
+        )
+
+        # Set the name and ID of this sensor to be the radiator name/id and a prefix.
+        super().__init__(
+            coordinator,
+            radiator,
+            name=f"{radiator.name} {description.name}",
+            unique_id=f"{radiator.id}-{description.key}",
+        )
+
+        self.entity_description = description
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the current sensor value (KW/h)."""
+        if self._radiator.energy_data:
+            return self._radiator.energy_data.kwh
+
+        return None
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Return the last time the sensor was initialized."""
+        if self._radiator.energy_data:
+            return self._radiator.energy_data.start
+
+        return None
+
+
+class RointePowerSensor(RointeRadiatorEntity, SensorEntity):
```

Comment:
These three sensors classes are very similar—I'm wondering if they could be combined into one (with the entity description providing what's necessary to display differences). Check out what I'm proposing for PurpleAir: https://github.com/home-assistant/core/blob/da72a38e383b4fb9071b8e9f1c1f61313fef0231/homeassistant/components/purpleair/sensor.py

Reply:
So, I stole your code from Purple Air :blush: which made the whole sensor platform much more elegant.

However I discovered a couple of snags: First I had to move the sensor description list to a different module, to avoid a circular import.

Secondly, I had to add a different method on the Coordinator to deal with the sensors. Basically my code isn't as sophisticated as PurpleAir's or Notion's.

I need to pass a list of Entity classes to the coordinator that then creates the new instance and populates the RointeEntity instance which is where all the data from the radiator is stored. 

On all the platforms used by this integration every entity contains a reference to `self._radiator` which is passed to the new entity on `add_entities_for_seen_keys` and now also on `add_sensor_entities_for_seen_keys`.

Unfortunately I don't have a proper mental model of how `ConfigEntry`, `async_setup_entry`, the DataCoordinator and `async_add_entities()` all should work together to proper refactor this. The current approach was (somehow painfully) put together by looking at other integrations and some comments I got on discord.

I leave to you to determine if you prefer this approach with a new method on the coordinator, the old approach, or something else. :) 

Additional optimizations:

- The `coordinator.add_entities_for_seen_keys()` method had a callback inside it which wasn't really necessary neither called from anywhere else. The methods was simplified.
- I noticed that only the coordinator was the only thing being stored on hass.data, so I removed the dictionary for a direct reference to the coordinator

Reply:
Got it. Let's not worry about it for now—there's nothing _wrong_ with these current classes other than some redundancy. If that will unstick you for now, that's fine; you can always return later if a new idea strikes you.

Reply:
If you approve I'll leave the current approach with an extra method to parse that sensor descriptions list. It's likely that new sensors will be added anyway.